lib/tree.rb

```diff
@@ -2,59 +2,168 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
+
+  def add(key, value)
+    if key == @key
+      return
+    elsif key < @key
+      left.nil? ? @left = TreeNode.new(key, value) : left.add(key, value)
+    elsif key > @key
+      right.nil? ? @right = TreeNode.new(key, value) : right.add(key, value)
+    end
+  end
+
+  def find(key)
+    if key == @key
+      return @value
+    elsif key < @key
+      left.nil? ? nil : left.find(key)
+    elsif key > @key
+      right.nil? ? nil : right.find(key)
+    end
+  end
+
+  def inorder(tree_array)
+    # if !left.nil?
+    if left
+      left.inorder(tree_array)
+    end
+
+    tree_array << { key: @key, value: @value }
+
+    # if !@right.nil?
+    if right
+      right.inorder(tree_array)
+    end
+  end
+
+  def preorder(tree_array)
+    tree_array << { key: @key, value: @value }
+
+    if left
+      left.preorder(tree_array)
+    end
+
+    if right
+      right.preorder(tree_array)
+    end
+  end
+
+  def postorder(tree_array)
+    if left
+      left.postorder(tree_array)
+    end
+
+    if right
+      right.postorder(tree_array)
+    end
+    tree_array << { key: @key, value: @value }
+  end
+
+  def height(tree_height)
+    # left_height = 0
+    # right_heigt = 0
+    # if !left.nil?
```

Comment:
I would say 

```ruby
unless left.nil?
  left_height = left.height + 1
end
```

You can do the same with right, and then return the larger of left and right's heights plus 1.